File: graphql-endpoint/src/main/java/nl/openweb/graphql_endpoint/kafka/CommandPublisher.java
package nl.openweb.graphql_endpoint.kafka;

import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.openweb.graphql_endpoint.properties.KafkaProperties;
import org.apache.avro.specific.SpecificRecordBase;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

@Slf4j
@AllArgsConstructor
@Service
public class CommandPublisher {

    private final KafkaProperties kafkaProperties;
    private final KafkaTemplate<String, Object> kafkaTemplate;

    public void publish(String username, SpecificRecordBase avroRecord) {
        kafkaTemplate.send(kafkaProperties.getCommandsTopic(), username, avroRecord);
    }
}

File: graphql-endpoint/src/main/kotlin/nl/openweb/graphql_endpoint/kafka/CommandPublisher.kt
package nl.openweb.graphql_endpoint.kafka

import io.micronaut.configuration.kafka.annotation.KafkaClient
import io.micronaut.configuration.kafka.annotation.KafkaKey
import io.micronaut.configuration.kafka.annotation.Topic
import org.apache.avro.specific.SpecificRecordBase

@KafkaClient("command-publisher")
interface CommandPublisher {

    @Topic("commands")
    fun publish(@KafkaKey username: String, avroRecord: SpecificRecordBase)
}